### PR TITLE
feat: Refactor usage of XHR for Config API

### PR DESCRIPTION
**Pull request by alexs-mparticle**
## Instructions
 1. PR target branch should be against `development`
 2. PR title name should follow this format: https://github.com/mParticle/mparticle-workflows/blob/main/.github/workflows/pr-title-check.yml
 3. PR branch prefix should follow this format: https://github.com/mParticle/mparticle-workflows/blob/main/.github/workflows/pr-branch-check-name.yml

 ## Summary
 - Replaces XHR Uploader with AsyncUploader, which allows config api requests to support both XHR and Fetch.

 ## Testing Plan
 - [x] Was this tested locally? If not, explain why.
 - Using both self-hosted and CDN sample apps, attempt to load the Web SDK and verify expected behavior.

 ## Reference Issue (For mParticle employees only.  Ignore if you are an outside contributor)
 - Closes https://go.mparticle.com/work/SQDSDKS-6566

### CODE REVIEW

File: src/identity.js [hunk 1]
```diff
@@ -1665,7 +1665,7 @@ export default function Identity(mpInstance) {
 
             mpInstance.Logger.verbose('Successfully parsed Identity Response');
 
-            mpInstance._APIClient.processQueuedEvents();
+            mpInstance._APIClient?.processQueuedEvents();
```
Comment (rmi22186) [inline src/identity.js:1668]:
```suggestion
            // https://go.mparticle.com/work/SQDSDKS-6568
            mpInstance._APIClient?.processQueuedEvents();
```

File: src/configAPIClient.ts [hunk 2]
```diff
@@ -91,78 +90,95 @@ export interface IConfigResponse {
...
-                apiKey +
-                '/config?env=';
-            if (config.isDevelopmentMode) {
-                url = url + '1';
-            } else {
-                url = url + '0';
-            }
+            const response = await uploader.upload(fetchPayload);
+            if (response.status === 200) {
+                console.log(
```
Comment (rmi22186) [inline src/configAPIClient.ts:150]: use mpinstance.logger instead of console.log

File: src/configAPIClient.ts [hunk 3]
```diff
@@ -91,78 +90,95 @@ export interface IConfigResponse {
...
-            const dataPlan = config.dataPlan as DataPlanConfig;
-            if (dataPlan) {
-                if (dataPlan.planId) {
-                    url = url + '&plan_id=' + dataPlan.planId || '';
-                }
-                if (dataPlan.planVersion) {
-                    url = url + '&plan_version=' + dataPlan.planVersion || '';
+                // https://go.mparticle.com/work/SQDSDKS-6568
+                // FetchUploader returns the response as a JSON object that we have to await
+                if (response.json) {
```
Comment (rmi22186) [inline src/configAPIClient.ts:159]: is this the only thing that will determine if it is a fetch uploader?  ether way, for readability, can we make the xhrresponse below part of an `else`?  it seems more clear that it's one or the other

Comment (rmi22186) [reply src/configAPIClient.ts:159]: then return the config response once to DRY it up.  i'm not committed to this idea, just a thought.

File: src/configAPIClient.ts [hunk 4]
```diff
@@ -91,78 +90,95 @@ export interface IConfigResponse {
...
+                // XHRUploader returns the response as a string that we need to parse
+                const xhrResponse = response as unknown as XMLHttpRequest;
+                configResponse = JSON.parse(xhrResponse.responseText);
+                return configResponse;
             }
+
+            mpInstance?.Logger?.verbose(
+                'Issue with receiving configuration from server, received HTTP Code of ' +
+                    response.statusText
+            );
```
Comment (rmi22186) [inline src/configAPIClient.ts:174]: i don't think this is ever reached?  I think this can be removed.

File: src/uploaders.ts [hunk 5]
```diff
@@ -33,26 +33,39 @@ export class XHRUploader extends AsyncUploader {
...
-            fetchPayload.body
+            fetchPayload.body,
+            fetchPayload.method as 'post' | 'get'
         );
         return response;
     }
 
-    private async makeRequest(url: string, data: string): Promise<Response> {
+    // XHR Ready States
+    // https://stackoverflow.com/questions/30522565/what-is-meaning-of-xhr-readystate-4#:~:text=State%204%20means%20that%20the,finished%20downloading%20the%20response%20content.
```
Comment (rmi22186) [inline src/uploaders.ts:43]: I'd rather use this link...just seems more reputable
```suggestion
    // https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/readyState
```

**Review by rmi22186 — CHANGES_REQUESTED** [review]
very minor nits

File: src/configAPIClient.ts (hunk 2 above)
Comment (alexs-mparticle) [reply src/configAPIClient.ts:150]: No need. There's already an mp logger below. This console.log was just for testing.

File: src/configAPIClient.ts (hunk 3 above)
Comment (alexs-mparticle) [reply src/configAPIClient.ts:159]: I'm good with that.

**Review by rmi22186 — APPROVED** [review]

**Pull request merged**